firedrake/extrusion_utils.py

```diff
@@ -43,119 +46,156 @@ def make_extruded_coords(extruded_topology, base_coords, ext_coords,
-        kernel = op2.Kernel("""
-static inline void pyop2_kernel_uniform_extrusion(double *ext_coords,
-                                                  const double *base_coords,
-                                                  const double *layer_height,
-                                                  int layer) {
-    for ( int d = 0; d < %(base_map_arity)d; d++ ) {
-        for ( int c = 0; c < %(base_coord_dim)d; c++ ) {
-            ext_coords[2*d*(%(base_coord_dim)d+1)+c] = base_coords[d*%(base_coord_dim)d+c];
-            ext_coords[(2*d+1)*(%(base_coord_dim)d+1)+c] = base_coords[d*%(base_coord_dim)d+c];
-        }
-        ext_coords[2*d*(%(base_coord_dim)d+1)+%(base_coord_dim)d] = *layer_height * (layer);
-        ext_coords[(2*d+1)*(%(base_coord_dim)d+1)+%(base_coord_dim)d] = *layer_height * (layer + 1);
-    }
-}""" % {'base_map_arity': base_coords.cell_node_map().arity, 'base_coord_dim': base_coords.function_space().value_size},
-            "pyop2_kernel_uniform_extrusion")
+        op2.ParLoop(kernel,
+                    ext_coords.cell_set,
+                    ext_coords.dat(op2.WRITE, ext_coords.cell_node_map()),
+                    base_coords.dat(op2.READ, base_coords.cell_node_map()),
+                    layer_height(op2.READ),
+                    pass_layer_arg=True,
+                    is_loopy_kernel=True).compute()
+        return
+    ext_fe = create_element(ext_coords.ufl_element())
+    ext_shape = ext_fe.index_shape
+    base_fe = create_element(base_coords.ufl_element())
+    base_shape = base_fe.index_shape
+    data = []
+    data.append(lp.GlobalArg("ext_coords", dtype=ScalarType, shape=ext_shape))
+    data.append(lp.GlobalArg("base_coords", dtype=ScalarType, shape=base_shape))
+    data.append(lp.GlobalArg("layer_height", dtype=ScalarType, shape=()))
+    data.append(lp.ValueArg('layer'))
+    base_coord_dim = base_coords.function_space().value_size
+    # Deal with tensor product cells
+    adim = len(ext_shape) - 2
+
+    def _get_arity_axis_inames(_base):
+        return tuple(_base + str(i) for i in range(adim))
+
+    def _get_lp_domains(_inames, _extents):
+        domains = []
+        for idx, extent in zip(_inames, _extents):
+            inames = isl.make_zero_and_vars([idx])
+            domains.append(((inames[0].le_set(inames[idx])) & (inames[idx].lt_set(inames[0] + extent))))
+        return domains
+
+    if extrusion_type == 'uniform':
+        domains = []
+        dd = _get_arity_axis_inames('d')
+        domains.extend(_get_lp_domains(dd, ext_shape[:adim]))
+        domains.extend(_get_lp_domains(('c', 'l'), (base_coord_dim, 2)))
+        instructions = """
+        <{IntType}> base_coord_dim = {bcdim}
+        ext_coords[{dd}, l, c] = base_coords[{dd}, c]
+        ext_coords[{dd}, l, base_coord_dim] = layer_height[0] * (layer + l)
+        """.format(IntType=IntType,
+                   dd=', '.join(dd),
+                   bcdim=base_coord_dim)
```

Comment:
Just hardcode `base_coord_dim` into the instructions directly, rather than declaring a temporary?

Reply:
Fixed.